### PR TITLE
Contributing lib/ansible/modules/network/cloudengine/ce_switchport.py module to manage HUAWEI data center CloudEngine

### PR DESCRIPTION
**Pull request by QijunPan**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

ce_switchport.py
##### ANSIBLE VERSION

```
ansible 2.3.0
config file = /etc/ansible/ansible.cfg
configured module search path = ['/home/ansible-2.1.1.0/lib/ansible/modules/core/network/cloudengine', '/home/ansible-2.1.1.0/lib/ansible/module_utils', '/home/ansible-2.1.1.0/lib/ansible/modules', '/home/ansible-2.1.1.0', '/home/ansible-2.1.1.0/lib/ansible/utils/']
```

##### SUMMARY

```
TASK [ENSURE 40GE1/0/4 is configured for access vlan 20] ***********************
task path: /usr1/code/OpenNESS/code/current/KVM/intg/test/testcases/test-ce_switchport.yml:30
Using module file /usr/local/lib/python2.7/site-packages/ansible-2.3.0-py2.7.egg/ansible/modules/core/network/cloudengine/ce_switchport.py
<ce_6850> ESTABLISH LOCAL CONNECTION FOR USER: root
<ce_6850> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1487053951.32-122707423167147 `" && echo ansible-tmp-1487053951.32-122707423167147="` echo $HOME/.ansible/tmp/ansible-tmp-1487053951.32-122707423167147 `" ) && sleep 0'
<ce_6850> PUT /tmp/tmpaG50mq TO /root/.ansible/tmp/ansible-tmp-1487053951.32-122707423167147/ce_switchport.py
<ce_6850> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1487053951.32-122707423167147/ /root/.ansible/tmp/ansible-tmp-1487053951.32-122707423167147/ce_switchport.py && sleep 0'
<ce_6850> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1487053951.32-122707423167147/ce_switchport.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1487053951.32-122707423167147/" > /dev/null 2>&1 && sleep 0'
changed: [ce_6850] => {
    "changed": true, 
    "end_state": {
        "access_vlan": "20", 
        "interface": "40GE1/0/4", 
        "mode": "access", 
        "native_vlan": "20", 
        "switchport": "enable", 
        "trunk_vlans": ""
    }, 
    "existing": {
        "access_vlan": "1", 
        "interface": "40GE1/0/4", 
        "mode": "access", 
        "native_vlan": "1", 
        "switchport": "enable", 
        "trunk_vlans": ""
    }, 
    "invocation": {
        "module_args": {
            "access_vlan": "20", 
            "auth_pass": null, 
            "authorize": false, 
            "host": "ce_6850", 
            "interface": "40GE1/0/4", 
            "mode": "access", 
            "native_vlan": null, 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 12345, 
            "provider": null, 
            "ssh_keyfile": null, 
            "state": "present", 
            "timeout": 10, 
            "transport": null, 
            "trunk_vlans": null, 
            "use_ssl": false, 
            "username": "rootDC", 
            "validate_certs": true
        }, 
        "module_name": "ce_switchport"
    }, 
    "proposed": {
        "access_vlan": "20", 
        "interface": "40GE1/0/4", 
        "mode": "access", 
        "native_vlan": null, 
        "state": "present", 
        "trunk_vlans": null
    }, 
    "updates": [
        "interface 40GE1/0/4", 
        "port default vlan 20"
    ]
}
```

### CODE REVIEW

File: lib/ansible/modules/network/cloudengine/ce_switchport.py [hunk 1]
```diff
@@ -0,0 +1,826 @@
...
+short_description: Manages Layer 2 switchport interfaces on HUAWEI CloudEngine switches.
+description:
+    - Manages Layer 2 switchport interfaces on HUAWEI CloudEngine switches.
+author: QijunPan (@CloudEngine-Ansible)
+notes:
+    - When C(state=absent), VLANs can be added/removed from trunk links and
+      the existing access VLAN can be 'unconfigured' to just having VLAN 1
+      on that interface.
+    - When working with trunks VLANs the keywords add/remove are always sent
+      in the `port trunk allow-pass vlan` command. Use verbose mode to see
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_switchport.py]: `C(port trunk allow-pass vlan)`

File: lib/ansible/modules/network/cloudengine/ce_switchport.py [hunk 2]
```diff
@@ -0,0 +1,826 @@
...
+            vlan_bit[num] = hex(tmp)[2]
+
+        vlan_xml = ''.join(vlan_bit)
+
+        return vlan_xml
+
+    def check_params(self):
+        """Check all input params"""
+
+        if not self.interface:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_switchport.py]: If you have `required=True` this isn't needed

File: lib/ansible/modules/network/cloudengine/ce_switchport.py [hunk 3]
```diff
@@ -0,0 +1,826 @@
...
+        iftype = 'stack-port'
+    elif interface.upper().startswith('NULL'):
+        iftype = 'null'
+    else:
+        return None
+
+    return iftype.lower()
+
+
+def is_portswitch_enalbe(iftype):
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_switchport.py]: is_portswitch_enalbe -> is_portswitch_enabled

File: lib/ansible/modules/network/cloudengine/ce_switchport.py [hunk 4]
```diff
@@ -0,0 +1,826 @@
...
+                msg='Error: Interface name must be set.')
+
+        # interface type check
+        if self.interface:
+            self.intf_type = get_interface_type(self.interface)
+            if not self.intf_type:
+                self.module.fail_json(
+                    msg='Error: Interface name of %s is error.' % self.interface)
+
+        if not self.intf_type or not is_portswitch_enalbe(self.intf_type):
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_switchport.py]: is_portswitch_enalbe -> is_portswitch_enabled

File: lib/ansible/modules/network/cloudengine/ce_switchport.py [hunk 5]
```diff
@@ -0,0 +1,826 @@
...
+            self.module.fail_json(msg='Error: Interface %s is error.')
+
+        # check access_vlan
+        if self.access_vlan:
+            if not self.access_vlan.isdigit():
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_switchport.py:707]: Not needed if you do `type='int'`

File: lib/ansible/modules/network/cloudengine/ce_switchport.py [hunk 6]
```diff
@@ -0,0 +1,826 @@
...
+                self.module.fail_json(msg='Error: Access vlan id is invalid.')
+            if int(self.access_vlan) <= 0 or int(self.access_vlan) > 4094:
+                self.module.fail_json(
+                    msg='Error: Access vlan id is not in the range from 1 to 4094.')
+
+        # check native_vlan
+        if self.native_vlan:
+            if not self.native_vlan.isdigit():
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_switchport.py:715]: Not needed if you do `type='int'`

File: lib/ansible/modules/network/cloudengine/ce_switchport.py [hunk 7]
```diff
@@ -0,0 +1,826 @@
...
+                self.module.fail_json(msg='Error: Native vlan id is invalid.')
+            if int(self.native_vlan) <= 0 or int(self.native_vlan) > 4094:
+                self.module.fail_json(
+                    msg='Error: Native vlan id is not in the range from 1 to 4094.')
+
+        # check mode
+        if self.state == "present" or self.state == "absent":
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_switchport.py]: Not needed if you have `required=True`

File: lib/ansible/modules/network/cloudengine/ce_switchport.py [hunk 8]
```diff
@@ -0,0 +1,826 @@
...
+
+        # present or absent
+        if self.state == "present" or self.state == "absent":
+            if self.mode == "access":
+                self.merge_access_vlan(self.interface, self.access_vlan)
+            elif self.mode == "trunk":
+                self.merge_trunk_vlan(
+                    self.interface, self.native_vlan, self.trunk_vlans)
+            else:
+                self.module.fail_json(msg='Error: Interface mode must be set.')
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_switchport.py]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
mutually_exclusive
required_together
required_one_of
require_if
```

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cloudengine/ce_switchport.py (hunk 6 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_switchport.py:715]: Let's keep this type, avoid conversion

File: lib/ansible/modules/network/cloudengine/ce_switchport.py (hunk 5 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_switchport.py:707]: Let's keep this type, avoid conversion

File: lib/ansible/modules/network/cloudengine/ce_switchport.py (hunk 4 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_switchport.py]: OK，thank you.

File: lib/ansible/modules/network/cloudengine/ce_switchport.py (hunk 8 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_switchport.py]: good idea,  thank you.

**Comment by QijunPan** [comment]
@gundalow   These problems have been fixed， can  this PR merged?

**Pull request merged**